xlators/mgmt/glusterd/src/glusterd-op-sm.c

```diff
@@ -3113,6 +3155,10 @@ glusterd_op_set_volume(dict_t *dict, char **errstr)
     gf_msg_debug(this->name, 0, "returning %d", ret);
     if (quorum_action)
         glusterd_do_quorum_action();
+    if (ret < 0 && count > 1)
+        glusterd_restore_volinfo_dict(volinfo, volinfo_dict_orig, dict);
```

Comment:
Instead of trying to sync `volinfo->dict` with `volinfo_dict_orig`, which may also fail in several points, wouldn't it be easier to always update `volinfo_dict_orig` and if everything works fine, we simply replace `volinfo->dict` with the new dict ? otherwise we destroy the new dict and return the error.

Reply:
Logically what you said makes sense (i.e.  work with the information received in the request dict and update the state information only if operation is successful).

But the problem is that the functions called by this function take only volinfo as the argument and at various places assume that the required info is present in volinfo->dict. 

So I think we would need to either (a) change function signature at various places to accept a new dict argument, or (b) add a new struct member glusterd_volinfo_t->temp_dict and use it in all functions and then  if success, update volinfo->dict finally. Both these seem a bit complex..

Reply:
Sample function backtrace just to indicate the call graph: 
```
#0  server_graph_builder (graph=0x7f46f8202bf0, volinfo=0x7f470001e2b0, set_dict=0x7f470001df58, param=0x7f470000bda0) at glusterd-volgen.c:2975
#1  0x00007f47084082b6 in build_graph_generic (graph=0x7f46f8202bf0, volinfo=0x7f470001e2b0, mod_dict=0x0, param=0x7f470000bda0, builder=0x7f470840d36c <server_graph_builder>) at glusterd-volgen.c:1086
#2  0x00007f470840d62a in build_server_graph (graph=0x7f46f8202bf0, volinfo=0x7f470001e2b0, mod_dict=0x0, brickinfo=0x7f470000bda0) at glusterd-volgen.c:3038
#3  0x00007f47084149d1 in glusterd_generate_brick_volfile (volinfo=0x7f470001e2b0, brickinfo=0x7f470000bda0, mod_dict=0x0, data=0x0) at glusterd-volgen.c:5803
#4  0x00007f47083b22a5 in _brick_for_each (volinfo=0x7f470001e2b0, mod_dict=0x0, data=0x0, fn=0x7f47084148c7 <glusterd_generate_brick_volfile>) at glusterd-utils.c:253
#5  0x00007f47083b2491 in glusterd_volume_brick_for_each (volinfo=0x7f470001e2b0, data=0x0, fn=0x7f47084148c7 <glusterd_generate_brick_volfile>) at glusterd-utils.c:300
#6  0x00007f4708415452 in generate_brick_volfiles (volinfo=0x7f470001e2b0) at glusterd-volgen.c:6031
#7  0x00007f470841774b in glusterd_create_volfiles (volinfo=0x7f470001e2b0) at glusterd-volgen.c:6761
#8  0x00007f4708417963 in glusterd_create_volfiles_and_notify_services (volinfo=0x7f470001e2b0) at glusterd-volgen.c:6805
#9  0x00007f47083a315f in glusterd_op_set_volume (dict=0x7f46f00036f8, errstr=0x7f46f8205dc8) at glusterd-op-sm.c:3096
```
The  custom xlator checks added in `server_graph_builder()` is what fails without this patch because stale values were present in the dict.

Reply:
Then we can do it in the reverse way: we copy the dict at the beginning but updates go to `volinfo->dict`. If everything succeeds, we are done. Otherwise we simply replace `volfinfo->dict` with the saved copy at the end of the function.

Reply:
But that is what the patch is doing, except instead of replacing the whole dict, it only the key-values modified in the set-volume command are restored in glusterd_restore_volinfo_dict() . @xhernandez Would you rather that we replace the entire dict instead? The only failures that can happen in the current glusterd_restore_volinfo_dict() function are ENOMEM related.

Reply:
I agree with xavi. I think in case of error we can simply reset volinfo->dict and call dict_copy with saved copy to replace the volinfo->dict.

Reply:
Done. I have force pushed the patch so that there is only 1 commit (instead of adding a function in 1 commit and removing it in the next commit).

Reply:
Sorry for the late answer. In this case replacing the dict is error free for sure and requires no computation at all (not taking mutexes nor searching). I think it's better, even if the chances of getting an error are very small.